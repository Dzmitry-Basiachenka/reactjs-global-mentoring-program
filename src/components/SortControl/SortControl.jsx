import React, { useState } from 'react';
import { Container, Form, Row, Col } from 'react-bootstrap';

const SortControl = ({ sortOptions, defaultSortBy, handleChange }) => {
    const [sortBy, setSortBy] = useState(defaultSortBy);

    const handleSortChange = (event) => {
        setSortBy(event.target.value);
        handleChange(event);
    }

    return (
        <Container>
            <Row>
                <Col md={3}>
                    <Form>
                        <Form.Group as={Row}>
                            <Form.Label column md={4}>
                                Sort by
                            </Form.Label>
                            <Col md={8}>
                                <Form.Select value={sortBy} onChange={handleSortChange}>
                                    {
                                        sortOptions.map((item) => (
                                            <option key={item.value} value={item.value}>
                                                {item.name}
                                            </option>
                                        ))
                                    }
                                </Form.Select>
                            </Col>
                        </Form.Group>
                    </Form>
                </Col>
            </Row>
        </Container>
    )
}

export default SortControl;
