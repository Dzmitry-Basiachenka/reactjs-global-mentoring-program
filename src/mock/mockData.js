export const MOCK_MOVIES = [
  {
    "id": 680,
    "title": "Pulp Fiction",
    "tagline": "Just because you are a character doesn't mean you have character.",
    "vote_average": 8.3,
    "vote_count": 10910,
    "release_date": "1994-09-10",
    "poster_path": "https://cdn.europosters.eu/image/1300/posters/pulp-fiction-cover-i1288.jpg",
    "overview": "A burger-loving hit man, his philosophical partner, a drug-addled gangster's moll and a washed-up boxer converge in this sprawling, comedic crime caper. Their adventures unfurl in three stories that ingeniously trip back and forth in time.",
    "budget": 8000000,
    "revenue": 213928762,
    "genres": [
      "Thriller",
      "Crime"
    ],
    "runtime": 154
  },
];

export const MOCK_SORT_OPTIONS = [
  {
    name: 'Release date',
    value: 'release_date'
  },
  {
    name: 'Title',
    value: 'title'
  },
];

export const MOCK_GENRES = [
  'Crime',
  'Documentary',
  'Horror',
  'Comedy'
];
