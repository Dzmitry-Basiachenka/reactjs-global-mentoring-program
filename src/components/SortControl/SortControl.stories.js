import SortControl from './SortControl.jsx';
import { MOCK_SORT_OPTIONS } from '../../mock/mockData.js';

export default {
  component: SortControl
};

export const Default = {
  args: {
    sortOptions: MOCK_SORT_OPTIONS,
    defaultSortBy: MOCK_SORT_OPTIONS[1].value,
    handleChange: event => console.log(`sort by: ${event.target.value}`),
  },
  parameters: {
  }
};
