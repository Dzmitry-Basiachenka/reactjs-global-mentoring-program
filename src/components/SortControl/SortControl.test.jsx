import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event';

import SortControl from './SortControl';
import { MOCK_SORT_OPTIONS } from '../../mock/mockData.js';

describe('SortControl', () => {

  afterEach(cleanup);

  it('should match snapshot', () => {
    const sortControl = renderer
      .create(
        <SortControl
          sortOptions={MOCK_SORT_OPTIONS}
          defaultSortBy={MOCK_SORT_OPTIONS[1].value}
        />
      )
      .toJSON();

    expect(sortControl).toMatchSnapshot();
  });

  it('render correctly', () => {
    render(
      <SortControl
        sortOptions={MOCK_SORT_OPTIONS}
        defaultSortBy={MOCK_SORT_OPTIONS[1].value}
      />
    );

    expect(screen.getByRole('combobox')).toHaveValue('title');
    expect(screen.getByRole('option', { name: 'Release date' }).selected).toBe(false);
    expect(screen.getByRole('option', { name: 'Title' }).selected).toBe(true);
  });

  it('should select value', async () => {
    const { user } = {
      user: userEvent.setup(),
      ...render(
        <SortControl
          sortOptions={MOCK_SORT_OPTIONS}
          defaultSortBy={MOCK_SORT_OPTIONS[1].value}
          handleChange={(event) => console.log(`sort by: ${event.target.value}`)}
        />
      )
    };

    expect(screen.getByRole('combobox')).toHaveValue('title');
    expect(screen.getByRole('option', { name: 'Release date' }).selected).toBe(false);
    expect(screen.getByRole('option', { name: 'Title' }).selected).toBe(true);

    await waitFor(async () => {
      await user.selectOptions(screen.getByRole('combobox'), 'release_date');
    });

    expect(screen.getByRole('combobox')).toHaveValue('release_date');
    expect(screen.getByRole('option', { name: 'Release date' }).selected).toBe(true);
    expect(screen.getByRole('option', { name: 'Title' }).selected).toBe(false);
  });

});
