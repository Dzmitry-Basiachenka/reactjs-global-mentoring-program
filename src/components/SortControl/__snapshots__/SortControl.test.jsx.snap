// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortControl should match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-md-3"
    >
      <form
        className=""
      >
        <div
          className="row"
        >
          <label
            className="form-label col-form-label col-md-4"
          >
            Sort by
          </label>
          <div
            className="col-md-8"
          >
            <select
              className="form-select"
              onChange={[Function]}
              value="title"
            >
              <option
                value="release_date"
              >
                Release date
              </option>
              <option
                value="title"
              >
                Title
              </option>
            </select>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
